### PR TITLE
Bring Enum.InternalBoxEnum to C# (CoreCLR)

### PR DESCRIPTION
**Pull request by Sergio0694**
This PR moves `Enum.InternalBoxValue` to C# in CoreCLR.
Opening as draft to double check all tests, the actual implementation seems to work fine locally.
Performance has a slight regression that I think is caused by the extra code having to call into `MethodTable.GetNumInstanceFieldBytes`. Should go away if we also move that to managed, which I think _should_ be doable?

|  Method | Branch |     Mean |    Error |   StdDev | Ratio | Code Size |
|-------- |------- |---------:|---------:|---------:|------:|----------:|
| EnumBox |   main | 69.01 ns | 1.013 ns | 0.948 ns |  1.00 |     379 B |
| EnumBox |     PR | **24.19 ns** | 0.284 ns | 0.266 ns |  **0.35** |     491 B |

<details>
    <summary><b>Benchmark code (click to expand):</b></summary>
<br/>

```csharp
using System.Reflection;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Running;

BenchmarkSwitcher.FromAssembly(Assembly.GetExecutingAssembly()).Run(args);

[DisassemblyDiagnoser]
[GroupBenchmarksBy(BenchmarkLogicalGroupRule.ByCategory)]
public class EnumBoxBenchmark
{
    [Benchmark]
    public object EnumBox()
    {
        return Enum.ToObject(typeof(Animal), 1);
    }
}

enum Animal
{
    Dog,
    Cat,
    Giraffe
}
```
</details>

### CODE REVIEW

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs [hunk 1]
```diff
@@ -663,4 +663,33 @@ internal unsafe struct TailCallTls
...
         public IntPtr ArgBuffer;
     }
+
+    // See src/vm/argslot.h
+    internal static class ArgSlot
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe byte* EndianessFixup(long* pArg, uint cbSize)
+        {
+#if BIGENDIAN
```
Comment (teo-tsirpanis) [inline src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:673]: I would use `if (!BitConverter.IsLittleEndian)` if I were you. In the BCL, the big-endian code gets trimmed away on little-endian platforms on compilation. Here's how ILSpy decompiles `BinaryPrimitives.ReadSingleBigEndian` for example:

image

Comment (Sergio0694) [reply src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:673]: That's interesting, just out of curiosity, why isn't the whole branch trimmed out? 🤔

Comment (teo-tsirpanis) [reply src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:673]: I guess it is important for the trimmer for some reason to keep referencing the field. It will almost certainly be optimized away by the JIT.

> Also, I figured using a directive would be better to minimize the IL size for the method?

I found 16 occurences of `#if !?BIGENDIAN` and 236 occurences of `if \(!?BitConverter.IsLittleEndian\)` in non-test C# code. The overhead of `ldsfld BitConverter.IsLittleEndian` and a branch is 7 bytes (plus a fixed cost for the field's metadata reference). 😅

File: src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs [hunk 2]
```diff
@@ -13,8 +13,35 @@ public abstract partial class Enum
...
+            // For more details, see src\vm\object.h.
+            // We need to get the type handle, so first we get a data reference, which gives us a
+            // reference to the start of the first data field (so m_keepalive), then shift
+            // by sizeof(void*) * 2. This gets us to the start of the m_typeHandle field.
+            ref byte dataRef = ref Unsafe.As<RawData>(enumType).Data;
+            ref byte typeHandleRef = ref Unsafe.AddByteOffset(ref dataRef, (nuint)(uint)sizeof(void*) * 2);
+            TypeHandle typeHandle = Unsafe.As<byte, TypeHandle>(ref typeHandleRef);
+
+            // The type here will never be a type desc, so we can just reinterpret
+            MethodTable* pMethodTable = typeHandle.AsMethodTable();
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:34]:
```suggestion
            MethodTable* pMethodTable = (MethodTable*)enumType.m_handle;
```

**Comment by jkotas** [comment]
> Performance has a slight regression 

You can make this faster by inlining the box:
- Allocate uninitialized object using `CreateUninitializedObject` (it may require a bit of work to make it callable from this place)
- switch over the InstanceBySize (there are only 4 possible sizes) and assign the value into the box directly.

The nice side-effect is that it will also get rid of the EndianessFixup that is not pretty.

**Comment by EgorBo** [comment]
> * Allocate uninitialized object using `CreateUninitializedObject`

I wonder if in this specific case it can use the GC's allocate with optional zeroing 🙂

**Comment by jkotas** [comment]
> I wonder if in this specific case it can use the GC's allocate with optional zeroing

GC's allocate with optional zeroing is for large buffers. Using it here would not help anything - quite the opposite.

File: src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs (hunk 2 above)
Comment (Sergio0694) [reply src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:34]: I'm not sure how I feel about the fact I did that before thinking of just doing this 😅

**Comment by Sergio0694** [comment]
> "switch over the InstanceBySize (there are only 4 possible sizes) and assign the value into the box directly."

Could you elaborate on what you mean by this? I'm not seeing an `InstanceBySize` method 🤔
Did you mean `GetNumInstanceFieldBytes`, or something else? I'm not sure I'm following that bit.

**Comment by Sergio0694** [comment]
I'm thinking it could also be nice to port `GetNumInstanceFieldBytes` to avoid the call into the runtime for that. It seems like it's just reading a bunch of fields but not doing anything particularly troubling otherwise. Porting `MethodTable::GetClass` seems pretty straightforward, the only tricky thing I can see is mapping `EEClass::m_cbBaseSizePadding` to the right offset in C# given the native object has a whole bunch of fields above it, including conditional ones for `DEBUG`/`COM_INTEROP`. Is there an easier way to get the base size padding from C#, or otherwise a way to calculate the right offset in a less error prone way than just manually calculating the offsets for all possible arch/debug/com combinations? 🤔

Thinking with that being managed too the whole method could be inlined, except for just the call to `JIT_New...` 😄
That `GetNumInstanceFieldBytes` is also used in other places, so it could give those a small perf bump as well.
Does this make sense?

**Comment by Sergio0694** [comment]
Also updated the benchmarks in the PR description, almost 3x faster already with the latest changes! 🚀

File: src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs [hunk 3]
```diff
@@ -13,8 +12,32 @@ public abstract partial class Enum
...
+                    break;
+                case 3:
+                    Unsafe.As<byte, uint>(ref dataRef) = (uint)value;
+                    break;
+                default:
+                    Unsafe.As<byte, ulong>(ref dataRef) = (ulong)value;
+                    break;
+            }
+
+            GC.KeepAlive(enumType);
```
Comment (teo-tsirpanis) [inline src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:37]: Can that line be moved after the call to `AllocateUninitializedObject`?

Comment (Sergio0694) [reply src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:37]: Do you mean relying on the fact the allocated object would also keep the same method table alive anyway? We still need that in the call to `GetNumInstanceFieldBytes` below that line. Not sure if moving this up would be worth it in terms of making the code more error prone - would leaving this here have much overhead at all? The GC would just keep the object potentially alive for a couple more instructions, but the call itself would still just go away, right? 🤔

Comment (teo-tsirpanis) [reply src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:37]: Yeah, the object holds the type, and the object is held because it will be returned. In either case it would make almost zero difference, this was mostly a cosmetic change.

File: src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs [hunk 4]
```diff
@@ -13,8 +12,32 @@ public abstract partial class Enum
...
+
+            switch (pMethodTable->GetNumInstanceFieldBytes())
+            {
+                case 1:
+                    dataRef = (byte)value;
+                    break;
+                case 2:
+                    Unsafe.As<byte, ushort>(ref dataRef) = (ushort)value;
+                    break;
+                case 3:
```
Comment (stephentoub) [inline src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:29]: 3?

Comment (Sergio0694) [reply src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs]: Whoops, small typo 😅
Fixed in https://github.com/dotnet/runtime/pull/70181/commits/483c9df45de4d3291f6322c66e5657c63341a648.

File: src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs [hunk 5]
```diff
@@ -13,8 +12,32 @@ public abstract partial class Enum
...
+                case 1:
+                    dataRef = (byte)value;
+                    break;
+                case 2:
+                    Unsafe.As<byte, ushort>(ref dataRef) = (ushort)value;
+                    break;
+                case 4:
+                    Unsafe.As<byte, uint>(ref dataRef) = (uint)value;
+                    break;
+                default:
```
Comment (stephentoub) [inline src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:32]:
```suggestion
                default:
                    Debug.Assert(pMethodTable->GetNumInstanceFieldBytes() == 8);
```
?

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs [hunk 6]
```diff
@@ -274,6 +274,14 @@ internal static unsafe bool ObjectHasComponentSize(object obj)
         [MethodImpl(MethodImplOptions.InternalCall)]
         internal static extern unsafe object? Box(MethodTable* methodTable, ref byte data);
 
+        /// <summary>
+        /// Allocates a new uninitialized object using an input <see cref="MethodTable"/> to determine its type.
+        /// </summary>
+        /// <param name="methodTable">The <see cref="MethodTable"/> pointer to use to create the instance.</param>
+        /// <returns>An instance of the type specified by <paramref name="methodTable"/>.</returns>
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        internal static extern unsafe object AllocateUninitializedObject(MethodTable* methodTable);
```
Comment (stephentoub) [inline src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:283]: Can GetUninitializedObject now be implemented in terms of AllocateUninitializedObject rather than having its own QCall and native impl?

Comment (Sergio0694) [reply src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:283]: It could, but it seems to me that would require some additional work:
- That API also has additional logic to validate the input type handle (in `RuntimeTypeHandle::ValidateTypeAbleToBeInstantiated`)
- Currently, the new `AllocateUninitializedObject` API will skip running static cctor (as that's not needed from the current callsite, as when an object is being cloned, the static cctor for that type has necessarily already been run). We'd need to port the logic to check that, so we might need a separate runtime call anyway.

We might potentially investigate that in a separate PR? 🙂

File: src/coreclr/vm/ecalllist.h [hunk 7]
```diff
@@ -598,6 +597,7 @@ FCFuncStart(gRuntimeHelpers)
     FCFuncElement("RegisterForGCReporting", GCReporting::Register)
     FCFuncElement("UnregisterForGCReporting", GCReporting::Unregister)
     FCFuncElement("Box", JIT_Box)
+    FCFuncElement("AllocateUninitializedObject", JIT_NewS_MP_FastPortable)
```
Comment (jkotas) [inline src/coreclr/vm/ecalllist.h:600]: This is correct in all cases. 
- There are special helpers for single-proc config. They have to be used when you are in the single proc config.
- arm32 needs to use a special helper that handles proper alignment.

There may be more corner cases that I am not thinking of at the moment.

I think you want to reuse the existing infra for allocating uninitialized objects as the other comment suggests.

File: src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs [hunk 8]
```diff
@@ -13,8 +13,33 @@ public abstract partial class Enum
...
+                    Debug.Assert(pMethodTable->GetNumInstanceFieldBytes() == 8);
+                    Unsafe.As<byte, ulong>(ref dataRef) = (ulong)value;
+                    break;
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:36]:
```suggestion
                case 8:
                    Unsafe.As<byte, ulong>(ref dataRef) = (ulong)value;
                    break;
                default:
                    Debug.Fail("Unexpected enum size");
                    break;
```

Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:36]: It is not worth it to try to save the extra branch

**Comment by jkotas** [comment]
> I'm thinking it could also be nice to port GetNumInstanceFieldBytes to avoid the call into the runtime for that.

Yes, it is possible, but I do not think it is worth. Accessing EEClass using FCalls is fine.

File: src/coreclr/vm/ecalllist.h (hunk 7 above)
Comment (Sergio0694) [reply src/coreclr/vm/ecalllist.h]: Ah, I see. Fixed in https://github.com/dotnet/runtime/pull/70181/commits/03b5b3315954c97c46dd431ce7c02cdaead92bdf with some guidance from Levi 🙂

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs [hunk 9]
```diff
@@ -3896,6 +3896,23 @@ private void CreateInstanceCheckThis()
...
 
+        /// <summary>
+        /// Creates an uninitialized object of the current type.
+        /// </summary>
+        /// <returns>An uninitalized instance of the object of the specified type.</returns>
+        [DebuggerStepThrough]
+        [DebuggerHidden]
+        internal object? CreateUninitializedInstance()
+        {
+            if (GenericCache is not ActivatorCache cache)
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:3907]: The GenericCache is used for enum names and values already. If there is a code that works with enums a lot, we will likely see the two cache uses fight between each other.

File: src/libraries/System.Private.CoreLib/src/System/Enum.EnumInfo.cs [hunk 10]
```diff
@@ -5,19 +5,47 @@ namespace System
...
+
+            /// <summary>
+            /// Creates an uninitialized object of the current type.
+            /// </summary>
+            /// <returns>An uninitalized instance of the object of the specified type.</returns>
+            internal object CreateUninitializedInstance(RuntimeType type)
+            {
+                Diagnostics.Debug.Assert(type.IsEnum);
+
+                if (_pfnAllocator is null)
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Enum.EnumInfo.cs:35]: This does not look thread-safe.

What guarantees that `_allocatorFirstArg` is initialized?

File: src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs [hunk 11]
```diff
@@ -13,8 +13,35 @@ public abstract partial class Enum
         [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Enum_GetValuesAndNames")]
         private static partial void GetEnumValuesAndNames(QCallTypeHandle enumType, ObjectHandleOnStack values, ObjectHandleOnStack names, Interop.BOOL getNames);
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern object InternalBoxEnum(RuntimeType enumType, long value);
+        private static unsafe object InternalBoxEnum(RuntimeType enumType, long value)
+        {
+            MethodTable* pMethodTable = (MethodTable*)enumType.m_handle;
+            object obj = GetEnumInfo(enumType, false).CreateUninitializedInstance(enumType);
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:19]: This is also going to fetch the enum values - it is more than what we need here.

File: src/libraries/System.Private.CoreLib/src/System/Enum.EnumInfo.cs (hunk 10 above)
Comment (Sergio0694) [reply src/libraries/System.Private.CoreLib/src/System/Enum.EnumInfo.cs:35]: It's not thread-safe, I just assumed in the worst case scenario the FCall would be repeated, but no harm done? From what I can see, other caching mechanisms for `RuntimeType` also do the same, with the cache potentially being initialized multiple times in case there's a race condition. I assume that that QCall is just returning the fixed address of the right JIT allocation helper, so there's no allocations involved and no concerns in case the call is made multiple times? 🤔

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Enum.EnumInfo.cs:35]: This is initializing two values: `_pfnAllocator` and `_allocatorFirstArg`. You want to make sure that both of these values are initialized before they are used below. This condition does not guarantee that both values are initialized.

Consider this interleaving:
Thread 1: Initializes `_pfnAllocator`
Thread 2: `if (_pfnAllocator is null)` condition is false, so the initialization is skipped
Thread 2: Executes `_pfnAllocator(_allocatorFirstArg);`. The allocator is going to crash since `_allocatorFirstArg` is not yet initalized.
Thread 1: Initializes `_allocatorFirstArg`.

File: src/libraries/System.Private.CoreLib/src/System/Enum.EnumInfo.cs [hunk 12]
```diff
@@ -5,19 +5,47 @@ namespace System
...
+        internal sealed unsafe class EnumInfo
         {
             public readonly bool HasFlagsAttribute;
             public readonly ulong[] Values;
             public readonly string[] Names;
 
+            // The cached managed function pointer to the right JIT allocation helper.
+            // Same as in RuntimeType.ActivationCache, just copied here to avoid cache conflicts.
+            private delegate*<void*, object> _pfnAllocator;
+            private void* _allocatorFirstArg;
```
[the remark this answers is not in the document]
Comment (Sergio0694) [reply src/libraries/System.Private.CoreLib/src/System/Enum.EnumInfo.cs]: Removed this as part of a6839324e49862f136c966a5f8b9069bff092a78.

File: src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs (hunk 11 above)
Comment (Sergio0694) [reply src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs]: I've reworked the logic in a6839324e49862f136c966a5f8b9069bff092a78 to only ever initialize the reflection-related info when needed, and otherwise only initialize the allocator info. Both are now initialized separately, but existing `EnumInfo` instances are otherwise reused. Hope this is better! 🙂

File: src/coreclr/System.Private.CoreLib/src/System/Enum.EnumInfo.CoreCLR.cs [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+                    GetEnumValuesAndNames(
+                        new QCallTypeHandle(ref enumTypeHandle),
+                        ObjectHandleOnStack.Create(ref values),
+                        ObjectHandleOnStack.Create(ref names),
+                        getNames ? Interop.BOOL.TRUE : Interop.BOOL.FALSE);
+
+                    bool hasFlagsAttribute = enumType.IsDefined(typeof(FlagsAttribute), inherit: false);
+
+                    HasFlagsAttribute = hasFlagsAttribute;
+                    Values = values!;
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Enum.EnumInfo.CoreCLR.cs:50]: This has race condition. It can overwrite already-initialized Names back to null.

Comment (Sergio0694) [reply src/coreclr/System.Private.CoreLib/src/System/Enum.EnumInfo.CoreCLR.cs:51]: Ah, right, I missed this corner case with a thread not requesting names racing against another one. I've added a compare exchange to fix this and a comment explaining the issue here. Left the previous fields lock-free since repeated writes should not cause issues as far as I can tell. Thanks! 🙂

File: src/libraries/System.Private.CoreLib/src/System/Enum.cs [hunk 14]
```diff
@@ -148,8 +148,8 @@ private static string ValueToHexString(object value)
 
         private static string? InternalFlagsFormat(EnumInfo enumInfo, ulong resultValue)
         {
-            string[] names = enumInfo.Names;
-            ulong[] values = enumInfo.Values;
+            string[] names = enumInfo.Names!;
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Enum.cs:151]: It would be nice to make Names and Values properties that automatically lazy-initialize the value so that these `!` are unnecessary.

It would also allow unification of `GetEnumInfo` and `GetEnumInfoForActivation` into one method. The split between  `GetEnumInfo` and `GetEnumInfoForActivation` is odd.

File: src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs [hunk 15]
```diff
@@ -13,35 +13,73 @@ public abstract partial class Enum
...
+        /// <param name="enumType">The enum type to use.</param>
+        /// <remarks>
+        /// This is like <see cref="GetEnumInfo(RuntimeType, bool)"/>, but skips unnecessary initialization.
+        /// </remarks>
+        private static EnumInfo GetEnumInfoForActivation(RuntimeType enumType)
         {
-            EnumInfo? entry = enumType.GenericCache as EnumInfo;
+            if (enumType.GenericCache is not EnumInfo entry)
+            {
+                enumType.GenericCache = entry = new EnumInfo();
```
Comment (deeprobin) [inline src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:63]: Has rather less to do with this PR. However, I wonder why EnumInfo is a class and not a struct.
Does EnumInfo have to be a reference type?

File: src/coreclr/System.Private.CoreLib/src/System/Enum.EnumInfo.CoreCLR.cs [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+                {
+                    ulong[]? values = null;
+                    string[]? names = null;
+                    RuntimeTypeHandle enumTypeHandle = enumType.TypeHandle;
+
+                    GetEnumValuesAndNames(
+                        new QCallTypeHandle(ref enumTypeHandle),
+                        ObjectHandleOnStack.Create(ref values),
+                        ObjectHandleOnStack.Create(ref names),
+                        getNames ? Interop.BOOL.TRUE : Interop.BOOL.FALSE);
```
Comment (deeprobin) [inline src/coreclr/System.Private.CoreLib/src/System/Enum.EnumInfo.CoreCLR.cs:46]: Technically, as far as I know, a bool can take more values than true/false. 
In that case, can we guarantee that we will only get true/false or whether the other side will accept any int?

If so, we could do the following:
```suggestion
                        Unsafe.As<bool, Interop.BOOL>(getNames);
```

File: src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj [hunk 17]
```diff
@@ -326,9 +327,9 @@
 
   <Target Name="GenerateEventingFiles"
           Inputs="@(EventingGenerationScript);@(EventManifestFile)"
-		  Outputs="@(EventingSourceFile)"
-		  DependsOnTargets="FindPython"
-		  BeforeTargets="BeforeCompile">
+          Outputs="@(EventingSourceFile)"
+          DependsOnTargets="FindPython"
+          BeforeTargets="BeforeCompile">
```
Comment (deeprobin) [inline src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj:332]: nit: Should be not part of this PR imo

**Review by deeprobin — APPROVED** [review]
LGTM

File: src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs [hunk 18]
```diff
@@ -13,35 +13,73 @@ public abstract partial class Enum
...
+                    break;
+                case 4:
+                    Unsafe.As<byte, uint>(ref dataRef) = (uint)value;
+                    break;
+                case 8:
+                    Unsafe.As<byte, ulong>(ref dataRef) = (ulong)value;
+                    break;
+                default:
+                    Debug.Fail("Unexpected enum size");
+                    break;
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:38]: Enums can't be backed by Int128.

File: src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs (hunk 15 above)
Comment (stephentoub) [reply src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs:63]: If it were a struct it would be boxed when storing into GenericCache.

File: src/coreclr/System.Private.CoreLib/src/System/Enum.EnumInfo.CoreCLR.cs (hunk 16 above)
Comment (stephentoub) [reply src/coreclr/System.Private.CoreLib/src/System/Enum.EnumInfo.CoreCLR.cs:46]: > `Unsafe.As<bool, Interop.BOOL>`

No, bool and Interop.BOOL are different sizes.

Comment (deeprobin) [reply src/coreclr/System.Private.CoreLib/src/System/Enum.EnumInfo.CoreCLR.cs:46]: Oh yes. I thought a `bool` is 4 big, but its only 1
